mcs/class/System/System.Net.Sockets/SafeSocketHandle.cs

```diff
@@ -55,20 +56,24 @@ protected override bool ReleaseHandle ()
 						break;
 					}
 
-					/*
-					 * This method can be called by the DangerousRelease inside RegisterForBlockingSyscall
-					 * When this happens blocking_threads contains the current thread.
-					 * We can safely close the socket and throw SocketException in RegisterForBlockingSyscall
-					 * before the blocking system call.
-					 */
 					lock (blocking_threads) {
+						/*
+						* This method can be called by the DangerousRelease inside RegisterForBlockingSyscall
+						* When this happens blocking_threads contains the current thread.
+						* We can safely close the socket and throw SocketException in RegisterForBlockingSyscall
+						* before the blocking system call.
+						*/
 						if (blocking_threads.Count == 1 && blocking_threads[0] == Thread.CurrentThread)
 							break;
-					}
 
-					AbortRegisteredThreads ();
-					// Sleep so other threads can resume
-					Thread.Sleep (1);
+						// abort registered threads
+						foreach (var t in blocking_threads)
+							Socket.cancel_blocking_socket_operation (t);
+
+						// Sleep so other threads can resume
+						in_cleanup = true;
+						Monitor.Wait (blocking_threads, 1000);
```

Comment:
Regarding https://github.com/mono/mono/pull/2779#discussion_r56983894

Conditional var waits must never be unconditional. They must always check the condition inside the lock before giving it up to Wait since a spurious wake up can happen.